### ISSUE TITLE
`debug_data` takes a lot of time doing nothing useful

### ISSUE THREAD

**Issue by bukzor** [HypothesisWorks/hypothesis-python#1040]
When I profile a hypothesis run (with use_coverage off), I see that 11% of the time is taken by `hypothesis.internal.conjecture.engine:398:debug_data`. Looking more closely, this function takes time generating data for the debugging log, even when that log is turned off. Let's please make this more lazy.

A very simple (but not elegant) solution is:

```diff
--- hypothesis/internal/conjecture/engine.py
+++ hypothesis/internal/conjecture/engine.py
@@ -26,6 +26,7 @@ from collections import defaultdict
 import attr
 
 from hypothesis import settings as Settings
+from hypothesis import Verbosity
 from hypothesis import Phase, HealthCheck
 from hypothesis.reporting import debug_report
 from hypothesis.internal.compat import EMPTY_BYTES, Counter, ceil, \
@@ -396,6 +397,8 @@ class ConjectureRunner(object):
             debug_report(message)
 
     def debug_data(self, data):
+        if self.settings.verbosity < Verbosity.debug:
+          return
         buffer_parts = [u"["]
         for i, (u, v) in enumerate(data.blocks):
             if i > 0:
```

**Comment by Zac-HD** [comment]
Hmm.  I've just traced this through, and we *could* pass a thunk to make string construction lazy.  I tend to think though that with only two calls it's more elegant just to check verbosity before calling the method.  I'll open a pull.

**Comment by bukzor** [comment]
Thanks! I'm eagerly awaiting.

**Comment by bukzor** [comment]
Thanks Zac!

On Jan 6, 2018 8:34 AM, "Zac Hatfield Dodds" <notifications@github.com>
wrote:

Closed #1040
<https://github.com/HypothesisWorks/hypothesis-python/issues/1040> via #1042
<https://github.com/HypothesisWorks/hypothesis-python/pull/1042>.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/HypothesisWorks/hypothesis-python/issues/1040#event-1412452419>,
or mute the thread
